api/lib/opentelemetry/resources/resource.rb

```diff
@@ -9,18 +9,41 @@ module Resources
     # Resource represents a resource, which captures identifying information about the entities
     # for which telemetry (metrics or traces) is reported.
     class Resource
+      # Returns the labels for this {Resource}
+      #
+      # @return [Hash<String, String>]
       attr_reader :labels
```

Comment:
The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-resources.md#retrieve-labels) says:

> The most common operation when retrieving labels is to enumerate over them. As such, it is recommended to optimize the resulting collection for fast enumeration over other considerations such as a way to quickly retrieve a value for a label with a specific key.

Should we expose an enumerator instead of a frozen hash?

Reply:
Just to clarify, as you suggesting that we return `@labels.to_enum`? If so, would we still call the method `labels`?

Reply:
I added a method `label_enumerator` that returns an enumerator, and kept the `labels` accessor, but changed its visibility to `protected` so that we can use it when merging.

Reply:
I'm not sure what exactly are we gaining here with `label_enumerator`. If I need to go over all the labels (which is the common case reading the spec) then `resource.label_enumerator.each {.` seems weird. I'd rather have `resource.labels.each` or just `resource.each` though the latter also seems a bit odd.

`@labels` is a hash which already is an enumerable. If the sdk implementor wants to create a different data structure for faster access then they just have to return a different enumerable from `def labels`.

I might be missing something here though

Reply:
The idea is to reduce the API surface area. Returning a frozen hash from `resource.labels` in the API sets the expectation that all implementations will have to return a frozen hash. The spec is more constrained - the expectation is that we support enumeration of labels but not indexing.

I agree that `resource.label_enumerator.each {` is weird. Renaming the method to `labels` and doing something else with the `attr_reader` is better. Or `resource.each_label {`?

Reply:
I'm open to any of these suggestions. The spec does allow us to:

> ...provide a way to retrieve a read only collection of labels associated with a resource

While returning a frozen hash of labels would force other SDKs to return something with the same interface, I'm not sure what other data structure they would use, and if it'd ever be a pain point to return a frozen hash.

From the user perspective, I could see wanting access to the underlying hash. For example, they may want to merge metadata together and the functions available on hash would be helpful.

If we think that would be a use case and that we'd ultimately end up exposing the hash, we should probably just expose the hash, and not the enumerator. On the other hand, if we think returning a hash would be a burden on SDK implementations, we may want to avoid it.

We can try to make a decision here in this PR, or we can defer that decision and open an issue. Thoughts @indrekj and @fbogsany?

Reply:
I'd prefer to defer the decision and open an issue.

Reply:
I opened #43. Should we merge this for the time being?